### PR TITLE
[vk] Extend pipeline cache functionalities

### PR DESCRIPTION
**Pull request by zakorgy**
Depends on https://github.com/gfx-rs/gfx/pull/2565

Added the `get_pipeline_cache_data` function to query pipeline cache data from Device. Added the optional `data` member to `create_pipeline_cache` where we can supply the bytes of a loaded cache.

PR checklist:
- [x] `make` succeeds (on *nix)
- [x] `make reftests` succeeds
- [x] tested examples with the following backends:
  - Vulkan (Windows, Linux)
- [ ] `rustfmt` run on changed code

---
This change is [Reviewable](https://reviewable.io/reviews/gfx-rs/gfx/2566)

### CODE REVIEW

File: src/backend/dx11/src/device.rs [hunk 1]
```diff
@@ -904,10 +904,15 @@ impl hal::Device<Backend> for Device {
         })
     }
 
-    fn create_pipeline_cache(&self) -> Result<(), device::OutOfMemory> {
+    fn create_pipeline_cache(&self, _data: Option<Vec<u8>>, device: &PhysicalDevice) -> Result<(), device::OutOfMemory> {
```
Comment (kvark) [inline src/backend/dx11/src/device.rs]: this should be `Option<&[u8]>`

Comment (kvark) [inline src/backend/dx11/src/device.rs]: why do we need the physical device here? It's not in https://www.khronos.org/registry/vulkan/specs/1.1-extensions/man/html/vkCreatePipelineCache.html

File: src/backend/vulkan/src/device.rs [hunk 2]
```diff
@@ -281,13 +281,70 @@ impl d::Device<B> for Device {
         }
     }
 
-    fn create_pipeline_cache(&self) -> Result<n::PipelineCache, d::OutOfMemory> {
+    fn create_pipeline_cache(
+        &self,
+        data: Option<Vec<u8>>,
+        device: &PhysicalDevice,
+    ) -> Result<n::PipelineCache, d::OutOfMemory> {
+        fn bad_cache(d: &[u8], properties: &vk::PhysicalDeviceProperties) -> bool {
```
Comment (kvark) [inline src/backend/vulkan/src/device.rs]: nit: let's reverse the logic and call it `is_valid` or something

File: src/backend/vulkan/src/device.rs [hunk 3]
```diff
@@ -281,13 +281,70 @@ impl d::Device<B> for Device {
...
+            if vendor_id != properties.vendor_id as u32 {
+                warn!(
+                    "Vendor ID mismatch. Device: {:?}, cache: {:?}.",
+                    properties.vendor_id,
+                    vendor_id,
+                );
+                return true;
+            }
+
+            let device_id = unsafe { std::mem::transmute::<[u8; 4], u32>([d[12], d[13], d[14], d[15]]) };
```
Comment (kvark) [inline src/backend/vulkan/src/device.rs]: looks like we should just cast the first 4 dwords right away, using `std::slice::from_raw_parts`

File: src/backend/dx11/src/device.rs (hunk 1 above)
Comment (kvark) [inline src/backend/dx11/src/device.rs]: these methods should be unsafe, just like all the other ones here

Comment (zakorgy) [reply src/backend/dx11/src/device.rs]: I couldn't find other way to access the `PhysicalDeviceProperties`  (for the cache validation) from the Device, so I placed the `&PhysicalDevice` here.
I also meant to propagate the validation to the create users, but they can't access backend specific resources e.g. `vk::PipelineCacheHeaderVersion` or `properties.pipeline_cache_uuid` from `gfx-hal`.
My other idea is to move the cache validation into the `PhysicalDevice` and the crate users can call it (e.g. `adpater.physical_device.is_valid_cache(&cache_bytes)`) before `create_pipeline_cache`.

Comment (kvark) [reply src/backend/dx11/src/device.rs]: We don't have to, strictly speaking, validate the caches on `create_pipeline_cache`, at least not at the API level. The backend can still keep VkPhysicalDevice around when creating a logical device and use it for invalidation.

File: src/backend/vulkan/src/lib.rs [hunk 4]
```diff
@@ -798,6 +798,50 @@ impl hal::PhysicalDevice<Backend> for PhysicalDevice {
             max_sampler_anisotropy: limits.max_sampler_anisotropy,
         }
     }
+
+    fn is_valid_cache(&self, cache: &[u8]) -> bool {
+        let cache_info: &[u32] = unsafe { slice::from_raw_parts(cache as *const _ as *const _, 4) };
```
Comment (kvark) [inline src/backend/vulkan/src/lib.rs]: need to assert/check here that the length is at least 16

File: src/backend/vulkan/src/lib.rs [hunk 5]
```diff
@@ -798,6 +798,50 @@ impl hal::PhysicalDevice<Backend> for PhysicalDevice {
...
+
+        if self.properties.pipeline_cache_uuid != cache[16..16 + vk::UUID_SIZE] {
+            warn!(
+                "Pipeline cache UUID mismatch. Device: {:?}, cache: {:?}.",
+                self.properties.pipeline_cache_uuid,
+                &cache[16..16 + vk::UUID_SIZE],
+            );
+            return false;
+        }
+        return true;
```
Comment (kvark) [inline src/backend/vulkan/src/lib.rs]: nit: no need for `return` here

**Pull request merged**